Convert docs from markdown to asciidoc #92

pyokagan opened the pull request:
Fixes #92 

```
Convert our documentation from markdown to asciidoc format, in order to
take advantage of the additional features and semantic elements that
asciidoc offers such as automatic section numbering, admonitions and
automatic table of contents (TOC) generation.

Implement a simple build+deployment system to automatically publish our
documentation to https://oss-generic.github.io/process/
```

---

The result (based on the tip of this PR) can be seen at https://pyokagan.github.io/oss-generic-process/
Please help to proof-read it! (Might make for some nice bedtime reading)

Note that the main goal of the PR is to convert the *bulk* documentation to asciidoc, so I'd rather not spend too much time on the build system aspect of this PR (e.g. using a full-featured static website generator or more advanced templating etc.) More advanced enhancements can be done later by someone else.

damithc commented:
Nice. Let me know when you expect some input from me.

CanIHasReview commented:
# v1

@pyokagan submitted v1 for review.

* [v1 1/15] [Translate markdown files into asciidoc files](https://github.com/oss-generic/process/pull/93/commits/83fc2e9376c583b59d8cdda23b41b8192b7b0813)
* [v1 2/15] [Setup asciidoc\-to\-html build system](https://github.com/oss-generic/process/pull/93/commits/cb71f0eea775c4bd986fada2518a6e8bbc097416)
* [v1 3/15] [build system: handle README\.adoc as an index page](https://github.com/oss-generic/process/pull/93/commits/98bbf2df2b47ebecde9feff76053df71171f2437)
* [v1 4/15] [build system: implement \`npm run watch\`](https://github.com/oss-generic/process/pull/93/commits/e08365ba2576a6facb204acb53b367820c4bdd80)
* [v1 5/15] [Use automatic TOC generation](https://github.com/oss-generic/process/pull/93/commits/d76838647d03c8b2ba9b5f65a806c391bcd36454)
* [v1 6/15] [Use inter\-document cross\-references](https://github.com/oss-generic/process/pull/93/commits/32cdbdbff22d700b23c6fd1b95d2c4879d6c6ebb)
* [v1 7/15] [Replace deep list nesting with subsections](https://github.com/oss-generic/process/pull/93/commits/c6bad426b44e429234a5755743fe1866627b68d6)
* [v1 8/15] [Restore HTML content erased by pandoc](https://github.com/oss-generic/process/pull/93/commits/b78dbcd1c0983e261ff38a25e2a1b8d96a7c57aa)
* [v1 9/15] [Fix "section title out of sequence" warnings](https://github.com/oss-generic/process/pull/93/commits/30aea3e0aaeedc2c69f1f82c623c2d693f53c2ea)
* [v1 10/15] [No explicit numbering of numbered list items](https://github.com/oss-generic/process/pull/93/commits/a0b6012d3dfb3391aa8a395a5486445e76fb33c3)
* [v1 11/15] [Use admonition/sidebar blocks instead of quote blocks](https://github.com/oss-generic/process/pull/93/commits/3711380072cb4f722a7b349e9f194be7d797a11b)
* [v1 12/15] [Remove unnecessary link anchors](https://github.com/oss-generic/process/pull/93/commits/43dc59ee3ca4388220f45c3a2aa302ecda13ec5b)
* [v1 13/15] [Fix rendering errors](https://github.com/oss-generic/process/pull/93/commits/4046452f1942945061f3f93cc862eea24242590e)
* [v1 14/15] [Remove markdown files](https://github.com/oss-generic/process/pull/93/commits/38a92df3eee5e04952b16ebc0c71468c79c7fc5a)
* [v1 15/15] [TO REMOVE BEFORE MERGE](https://github.com/oss-generic/process/pull/93/commits/40e2f16eb5a591d9939d0933e0e40df4fa3c4e1c)

([:books: Archive](https://github.com/oss-generic/process/compare/755ceae589f2935540ef229dc9e99f2b257ed062...40e2f16eb5a591d9939d0933e0e40df4fa3c4e1c))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/oss-generic/process.git refs/pr/93/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

pyokagan commented:
+cc documentation authors @Skaty @li-kai  @AngShiYa @chao1995 @yamgent @acjh

damithc commented on build.js:
```diff
@@ -0,0 +1,181 @@
+function onError(err) {
+    console.error(err);
+    process.exit(1);
+}
```
Did we write this file ourselves or got it from somewhere?

pyokagan replied:
I wrote it.

damithc commented on docs/AdvancedContributorGuidelines.adoc:
```diff
@@ -0,0 +1,13 @@
+== Updating a PR in response to a review
+
+* Favor the <<HowToGuides.adoc#create-a-multi-step-pr,multi-step PR approach>> even for smaller PRs which has only one
+logical step.
```
The link points to a non-existent file. Remove the whole point?

damithc commented on docs/FormatsAndConventions.adoc:
```diff
@@ -0,0 +1,307 @@
+For example, the commit message title
+`AbstractPerson: remove add(int) method` is acceptable in place of
+`AbstractPerson: remove AbstractPerson#add(int) method`.
+====
```
Perhaps we can use this 'box style' for all examples everywhere. Probably out of scope for this PR.

pyokagan replied:
> Probably out of scope for this PR.

Yeah, I'll leave this for someone else to do. (This commit just deals with pre-existing blockquote blocks)

damithc commented:
I forgot to add: Some of the the `// TODO ` fillers in the previous version seem to have disappeared, resulting in section titles without any body.

pyokagan commented:
@damithc 
> Some of the the // TODO fillers in the previous version seem to have disappeared, resulting in section titles without any body.

OK.

CanIHasReview commented:
# v2

@pyokagan submitted v2 for review.

* [v2 1/17] [Translate markdown files into asciidoc files](https://github.com/oss-generic/process/pull/93/commits/5327d406a3cf7140762e7ed60a469f6198805fde)
* [v2 2/17] [Setup asciidoc\-to\-html build system](https://github.com/oss-generic/process/pull/93/commits/443585b689428c53e293683805de426c1bc19a05)
* [v2 3/17] [build system: handle README\.adoc as an index page](https://github.com/oss-generic/process/pull/93/commits/29fcea52e5f72ae55610268b6b93f44953c30ff8)
* [v2 4/17] [Use automatic TOC generation](https://github.com/oss-generic/process/pull/93/commits/ae5aebcd2538aabb608d5ee01f159e9410503f91)
* [v2 5/17] [Use inter\-document cross\-references](https://github.com/oss-generic/process/pull/93/commits/e19e5591ce41cafc04cdfd15d5939cb57eccc510)
* [v2 6/17] [Replace deep list nesting with subsections](https://github.com/oss-generic/process/pull/93/commits/38819e5b87cadf838fedfddccbd67fc61e7f4543)
* [v2 7/17] [Restore HTML content erased by pandoc](https://github.com/oss-generic/process/pull/93/commits/7ff82d1b4f4f51303ad700583fd50ea3a6caa05c)
* [v2 8/17] [Fix "section title out of sequence" warnings](https://github.com/oss-generic/process/pull/93/commits/540677fcf84b022e0023b1e859692db69d5c05dc)
* [v2 9/17] [Remove explicit numbering of numbered list items](https://github.com/oss-generic/process/pull/93/commits/e9dd6489f3f0ddbae36fec420eec2b1ac46954e1)
* [v2 10/17] [Use admonition/sidebar blocks instead of quote blocks](https://github.com/oss-generic/process/pull/93/commits/3e33a3e685f64c28c66fd70725899bc87d61b29d)
* [v2 11/17] [Remove unnecessary link anchors](https://github.com/oss-generic/process/pull/93/commits/8af5afd98df3749d822935feeaa4d5f286921ef1)
* [v2 12/17] [Ensure asciidoctor renders certain lines starting with \`//\`](https://github.com/oss-generic/process/pull/93/commits/605eedf44ee9ada39eeda8eb1b5f014163e1cefc)
* [v2 13/17] [Fix rendering errors in all adoc documents](https://github.com/oss-generic/process/pull/93/commits/2beb6b93f90a56f0267977d0238f53dcf50f7ef4)
* [v2 14/17] [Remove markdown files](https://github.com/oss-generic/process/pull/93/commits/3bb2781256ca836a312c10003ea4adb9be21efdf)
* [v2 15/17] [AdvancedContributorGuidelines: remove "Updating a PR in response to a review"](https://github.com/oss-generic/process/pull/93/commits/3bd5bfd300be081cb9f4331082a401d786b1ded7)
* [v2 16/17] [CodingStandard\-Java\.adoc: introduce and use \{star\} substitution](https://github.com/oss-generic/process/pull/93/commits/9b230e156eb91949754e349a4ca9c2a380c6515c)
* [v2 17/17] [TO REMOVE BEFORE MERGE](https://github.com/oss-generic/process/pull/93/commits/76ce70c14a0aed12831f28a191aea6bfff1e583c)

([:books: Archive](https://github.com/oss-generic/process/compare/386d8b79c5bc592f31b3e05b96d0ee35585c68e5...76ce70c14a0aed12831f28a191aea6bfff1e583c)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/oss-generic/process/compare/5fd26f821cb36cabc99b22a51035d1f62cbf061d~1...5fd26f821cb36cabc99b22a51035d1f62cbf061d))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/oss-generic/process.git refs/pr/93/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

pyokagan commented:
v2

* Simplified `build.js` script.
* Rebased on master, including the massive changes in the Java coding standard. (I actually redid all the commits, but with an updated pandoc version which generated slightly different asciidoc output, which is why there are random changes in the interdiff).

TIL: `:trollface:` isn't an actual unicode emoji :trollface: I used an image instead.

pyokagan commented:
@damithc Done. You'll need to set up Travis CI on this repo with the `GITHUB_TOKEN` environment variable set to a personal access token of an account which can push to this repo. The usual stuff :-)
